// playwright/lib/cli/client/registry.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var registry_exports = {};
__export(registry_exports, {
  Registry: () => Registry,
  baseDaemonDir: () => baseDaemonDir,
  createClientInfo: () => createClientInfo
});
module.exports = __toCommonJS(registry_exports);
var import_crypto = __toESM(require("crypto"));
var import_fs = __toESM(require("fs"));
var import_os = __toESM(require("os"));
var import_path = __toESM(require("path"));
class Registry {
  constructor(entries) {
    this._entries = entries;
  }
  entry(clientInfo, sessionName) {
    const key = clientInfo.workspaceDir || clientInfo.workspaceDirHash;
    const entries = this._entries.get(key) || [];
    return entries.find((entry) => entry.config.name === sessionName);
  }
  entries(clientInfo) {
    const key = clientInfo.workspaceDir || clientInfo.workspaceDirHash;
    return this._entries.get(key) || [];
  }
  entryMap() {
    return this._entries;
  }
  static async loadSessionEntry(file) {
    try {
      const data = await import_fs.default.promises.readFile(file, "utf-8");
      const config = JSON.parse(data);
      if (!config.name)
        config.name = import_path.default.basename(file, ".session");
      if (!config.timestamp)
        config.timestamp = 0;
      return { file, config };
    } catch {
      return void 0;
    }
  }
  static async load() {
    const sessions = /* @__PURE__ */ new Map();
    const hashDirs = await import_fs.default.promises.readdir(baseDaemonDir).catch(() => []);
    for (const workspaceDirHash of hashDirs) {
      const hashDir = import_path.default.join(baseDaemonDir, workspaceDirHash);
      const stat = await import_fs.default.promises.stat(hashDir);
      if (!stat.isDirectory())
        continue;
      const files = await import_fs.default.promises.readdir(hashDir).catch(() => []);
      for (const file of files) {
        if (!file.endsWith(".session"))
          continue;
        const fileName = import_path.default.join(hashDir, file);
        const entry = await Registry.loadSessionEntry(fileName);
        if (!entry)
          continue;
        const key = entry.config.workspaceDir || workspaceDirHash;
        let list = sessions.get(key);
        if (!list) {
          list = [];
          sessions.set(key, list);
        }
        list.push(entry);
      }
    }
    return new Registry(sessions);
  }
}
const baseDaemonDir = (() => {
  if (process.env.PLAYWRIGHT_DAEMON_SESSION_DIR)
    return process.env.PLAYWRIGHT_DAEMON_SESSION_DIR;
  let localCacheDir;
  if (process.platform === "linux")
    localCacheDir = process.env.XDG_CACHE_HOME || import_path.default.join(import_os.default.homedir(), ".cache");
  if (process.platform === "darwin")
    localCacheDir = import_path.default.join(import_os.default.homedir(), "Library", "Caches");
  if (process.platform === "win32")
    localCacheDir = process.env.LOCALAPPDATA || import_path.default.join(import_os.default.homedir(), "AppData", "Local");
  if (!localCacheDir)
    throw new Error("Unsupported platform: " + process.platform);
  return import_path.default.join(localCacheDir, "ms-playwright", "daemon");
})();
function createClientInfo() {
  const packageLocation = require.resolve("../../../package.json");
  const packageJSON = require(packageLocation);
  const workspaceDir = findWorkspaceDir(process.cwd());
  const version = process.env.PLAYWRIGHT_CLI_VERSION_FOR_TEST || packageJSON.version;
  const hash = import_crypto.default.createHash("sha1");
  hash.update(workspaceDir || packageLocation);
  const workspaceDirHash = hash.digest("hex").substring(0, 16);
  return {
    version,
    workspaceDir,
    workspaceDirHash,
    daemonProfilesDir: daemonProfilesDir(workspaceDirHash)
  };
}
function findWorkspaceDir(startDir) {
  let dir = startDir;
  for (let i = 0; i < 10; i++) {
    if (import_fs.default.existsSync(import_path.default.join(dir, ".playwright")))
      return dir;
    const parentDir = import_path.default.dirname(dir);
    if (parentDir === dir)
      break;
    dir = parentDir;
  }
  return void 0;
}
const daemonProfilesDir = (workspaceDirHash) => {
  return import_path.default.join(baseDaemonDir, workspaceDirHash);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Registry,
  baseDaemonDir,
  createClientInfo
});
